doc/static-route/SONiC_static_route_bfd_hld.md

```diff
@@ -0,0 +1,300 @@
+3. Install, update, removing static route (i.e., update nexthop, remove static route if no nexthop) based on BFD session state change.<br>
+4. When the application (static route BFD) restarts, recover static routes and BFD session states from redis database (config_db, appl_db and state_db) without impacting any existing installed static routes and BFD sessions. <br  />  
+
+## System Overview with Static Route BFD
+A new component, StaticRouteBfd is introduced to support static route BFD. <br>
+
+In the following diagram, StaticRouteBfd monitors config_db STATIC_ROUTE_TABLE. If there is a static route configured with bfd, StaticRouteBfd creates BFD session by updating appl_db BFD_SESSION_TABLE. StaticRouteBfd also monitors state db for BFD session state change. Based on BFD session state, StaticRouteBfd updates appl_db STATIC_ROUTE_TABLE to add/delete static route or update its nexthop list.<br>
+
+To work with existing bgpcfgd StaticRouteMgr, an optional field "bfd" is introduced in STATIC_ROUTE_TABLE. When the "bfd" field present and is "true" in config db, StaticRouteMgr ignores this static route and StaticRouteBfd handles it. StaticRouteBfd writes this static route to appl_db based on BFD session state for this static route nexthop. <br>
+
+To avoid StaticRouteTimer deleting StaticRouteBfd created static route entry in appl_db, a new field "expiry" is introduced in appl_db STATIC_ROUTE_TABLE schema. StaticRouteBfd sets "expiry"="false" when it writes a static route to appl_db STATIC_ROUTE_TABLE. When StaticRouteTimer see "expiry"="false", it ignores this static route entry. <br>
+
+
+<img src="static_rt_bfd_overview.png" width="500">
+<br />
+
+## BFD session local IP address
+StaticRouteBfd create BFD session for each nexthop inside a BFD enabled static route. The BFD peer IP is the nexthop, the BFD local address (source IP address of the BFD packet) is the lookup result using the following method:<br />
+
+1. find the interface name for the nexthop
+2. lookup the config_db, find the IP address of that interface
+3. choose the IP address which has same type of nexthop (IPv4 or IPv6)
+4. If no IP address found, use local loopback IP address, and log a warning message for that. <br /><br />
+
+In the example (assume the "bfd" field is "true"), for the nexthop "20.0.10.3", the corresponding  interface name is PortChannel10.
+From the interface configuration, we can found two IP addresses, an IPv4 address 20.0.10.1 and an IPv6 address 2603:10E2:400:10::1. we choose 20.0.10.1 because the nexthop IP address is IPv4 address. <br /><br />
+
+<img src="static_rt_bfd_cfg.png" width="800">
+<img src="static_rt_bfd_interface.png" width="500">
+<br />
+
+## DB changes
+ 
+Two optional fields are introduced to STATIC_ROUTE_TABLE: <br>
+1. "bfd" -- Default value is false when this field is not configured
+2. "expiry" -- Default value is true when this field is not configured
```

Comment:
Not sure why we need this "expiry" field, as we expect to have only non-persistent STATIC_ROUTE_TABLE in the APPL_DB.

Reply:
related code:
https://github.com/sonic-net/sonic-buildimage/blob/bce824723c0d2c5d9cb7ae34bb830bb632d43c83/src/sonic-bgpcfgd/bgpcfgd/static_rt_timer.py#L36
to let static_rt_timer not remove the static route.

Reply:
I think this code is being used only for non-persistent route but in this HLD we focus only persistent route (CONFIG_DB) which means there should not be any STATIC_ROUTE_TABLE in the APPL_DB or in other words, we should not worry about STATIC_ROUTE_TABLE presence in the APPL_DB, do we use STATIC_ROUTE_TABLE present in APPL_DB for something?

I thought, we could send only the BFD_SESSION information all the way to SAI via APPL_DB & ASIC_DB and get back the notification if the session is down/up for us to react in the StaticRouteBfd module.

Reply:
The process is that, StaticRouteBfd read static route from config_db (persistent), and write the processed route to appl_db STATIC_ROUTE_TABLE, to utilize StaticRouteMgr(appl_db) to install the route to FRR. so the processed route looks exact same as non-persistent route (to StaticRouteMgr(appl_db)). that is the reason we set expiry=false to avoid StaticRouteTimer deleting it.

for your last comments, the bfdorch is already in the system to provide bfd service, write request to appl_db BFD_SESSION_TABLE can utilize the current bfdorch service. bfdorch maintains local_discriminator and BFD sessions in sonic system, other service should not create BFD session by directly calling SAI API.